### PR TITLE
fixed issue with .has() when new value is added and then deleted or c…

### PR DESCRIPTION
**Pull request by overthemike**
…leared

## Related Bug Reports or Discussions
#979
Fixes #
#979 
## Summary
Accessing this.index in the has method doesn't seem to work when a new item is added and then deleted (or cleared) at a later time. Accessing the this.data.length property fixes this issue.

## Check List

- [ x ] `pnpm run prettier` for formatting code and docs

### CODE REVIEW

File: src/vanilla/utils/proxyMap.ts
Comment (dai-shi) [inline src/vanilla/utils/proxyMap.ts:93]: Let's do this:
```suggestion
      return map.has(key)
```

**Review by dai-shi — COMMENTED** [review]
Thanks for working on it. Is it possible to add a test?

File: src/vanilla/utils/proxyMap.ts
Comment (overthemike) [reply src/vanilla/utils/proxyMap.ts]: Yeah I was planning on it actually. Just thought I'd push this up first to get the patch out. I didn't have a whole lot of time today with other work. Do you want me to push up a different pr for the tests or do you want it in this one?

Comment (overthemike) [reply src/vanilla/utils/proxyMap.ts]: I added the tests and pushed up

File: src/vanilla/utils/proxyMap.ts
Comment (dai-shi) [inline src/vanilla/utils/proxyMap.ts:91]: What is the case `this.data.length` changes, but `this.index` does not? It feels like we need another number like `epoch`.

File: tests/proxyMap.test.tsx
Comment (dai-shi) [inline tests/proxyMap.test.tsx:389]: Try something like this:
```suggestion
          <button onClick={() => { state.set('key', 'value'); state.set('key2', 'value'); }}>set key</button>
```

**Review by dai-shi — CHANGES_REQUESTED** [review]
it probably requires more.

File: src/vanilla/utils/proxyMap.ts
Comment (overthemike) [reply src/vanilla/utils/proxyMap.ts:91]: The case I put in the tests was precisely the case that didn't work with this.index. When I wrote the tests, I went back to this.index just to make sure I wasn't missing something and the test didn't pass. When a key is deleted or cleared, `.has()` wasn't updating the ui with `useSnapshot`. The tests we had before was just testing snapshot, which was taken manually.

Out of curiosity, is there a reason you don't like checking this.data.length? It's O(1).

File: tests/proxyMap.test.tsx
Comment (overthemike) [reply tests/proxyMap.test.tsx:389]: This is a good reason 😅😬

Comment (overthemike) [reply tests/proxyMap.test.tsx:389]: I added an epoch to keep track of versions instead of using this.data.length and added tests to test for adding and deleting multiple keys as well as clearing the map/set

File: src/vanilla/utils/proxyMap.ts
Comment (dai-shi) [reply src/vanilla/utils/proxyMap.ts]: `this.data.length` doesn't change if we delete an item (maybe except for the last one.)

Comment (overthemike) [reply src/vanilla/utils/proxyMap.ts]: I originally thought that in order for updates to happen, `this.data.length` had to have changed, but I'm not sure if that's the case. I'm not really sure why it (sort of) works when deleting just one item. I think it's because we're tracking touch the property? I also don't really understand why it would work with a single element, but not with multiple elements. Might only update for the first one. Either way, epoch works with both, and I think it's probably easier to understand what it's doing rather than just accessing the length property randomly on an array....👍😁

**Comment by overthemike** [comment]
Is there anything else that needs doing before this gets merged?

File: src/vanilla/utils/proxyMap.ts
Comment (dai-shi) [reply src/vanilla/utils/proxyMap.ts]: It's because we use a sparse array.

File: src/vanilla/utils/proxyMap.ts
Comment (dai-shi) [inline src/vanilla/utils/proxyMap.ts:94]:
```suggestion
      this.epoch // touch property for tracking
```

File: tests/proxyMap.test.tsx
Comment (dai-shi) [inline tests/proxyMap.test.tsx:424]: I think we should have a test that deletes the first key only. (Did I miss it somewhere?)

**Review by dai-shi — COMMENTED** [review]
Looks good overall.

File: tests/proxyMap.test.tsx
Comment (overthemike) [reply tests/proxyMap.test.tsx:424]: I'll add it

Comment (overthemike) [reply tests/proxyMap.test.tsx:424]: I went ahead and added another test that tests deleting the first key as well as deleting only the second (or last) key. Couldn't hurt.

**Review by dai-shi — APPROVED** [review]
This causes extra rerenders with `.has()`, which doesn't seem ideal, but can't be helped. It's ironic that the use of `.has()` should be discouraged, even though proxyMap/Set should exist for performance.

**Comment by overthemike** [comment]
Yeah - `.has()` essentially needs to check if the value exists every time the data changes :\

**Comment by overthemike** [comment]
Are we waiting on anything to get this merged and get a patch version published?

**Comment by dai-shi** [comment]
https://github.com/pmndrs/valtio/releases/tag/v2.1.1

**Pull request merged**
